src/inmanta/data/__init__.py

```diff
@@ -1316,6 +1321,28 @@ async def get(cls, env: uuid.UUID, endpoint: str, connection: Optional[asyncpg.c
         obj = await cls.get_one(environment=env, name=endpoint, connection=connection)
         return obj
 
+    @classmethod
+    async def persist_on_halt(cls, env: uuid.UUID) -> None:
+        """
+            Persists paused state when halting all agents.
+        """
+        await cls._execute_query(
+            f"UPDATE {cls.table_name()} SET unpause_on_resume=NOT paused WHERE environment=$1", cls._get_value(env)
+        )
+
+    @classmethod
+    async def persist_on_resume(cls, env: uuid.UUID) -> List[str]:
+        """
+            Restores default halted state. Returns a list of agents that should be unpaused.
+        """
+        unpause_on_resume = await cls._fetch_query(
+            f"SELECT name FROM {cls.table_name()} WHERE environment=$1 AND unpause_on_resume", cls._get_value(env)
+        )
+        await cls._execute_query(
+            f"UPDATE {cls.table_name()} SET unpause_on_resume=null WHERE environment=$1", cls._get_value(env)
```

Comment:
This could perhaps be done using UPDATE RETURNING?

Reply:
I don't think so because `UPDATE RETURNING` returns from the database after the update if I understand correctly.

Reply:
In that case, I would make sure this runs in one transaction

Reply:
You could do something like this: `UPDATE {cls.table_name()} SET unpause_on_resume=null WHERE environment=$1 AND unpause_on_resume RETURNING name", cls._get_value(env)`. The where clause will make sure you only return the records for which unpause_on_resume was true.

Reply:
Indeed. It's probably cleaner to make sure that everything is set to null in the end.

Reply:
I agree. I had thought about doing it like that but it would still require a second query to set the others to `null`. I'll wrap it in a transaction.